File: AirshipKit/AirshipKit/UAProximityRegion.m
/*
 Copyright 2009-2016 Urban Airship Inc. All rights reserved.

 Redistribution and use in source and binary forms, with or without
 modification, are permitted provided that the following conditions are met:

 1. Redistributions of source code must retain the above copyright notice, this
 list of conditions and the following disclaimer.

 2. Redistributions in binary form must reproduce the above copyright notice,
 this list of conditions and the following disclaimer in the documentation
 and/or other materials provided with the distribution.

 THIS SOFTWARE IS PROVIDED BY THE URBAN AIRSHIP INC ``AS IS'' AND ANY EXPRESS OR
 IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF
 MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO
 EVENT SHALL URBAN AIRSHIP INC OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT,
 INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
 BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
 DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
 LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE
 OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF
 ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */

#import "UAProximityRegion+Internal.h"
#import "UARegionEvent+Internal.h"
#import "UAGlobal.h"

@implementation UAProximityRegion

+ (instancetype)proximityRegionWithID:(NSString *)proximityID major:(NSNumber *)major minor:(NSNumber *)minor {

    UAProximityRegion *proximityRegion = [[self alloc] init];

    proximityRegion.major = major;
    proximityRegion.minor = minor;
    proximityRegion.proximityID = proximityID;

    if (!proximityRegion.isValid) {
        return nil;
    }

    return proximityRegion;
}

- (void)setLatitude:(NSNumber *)latitude {
    if (latitude != _latitude) {

        if (latitude && ![UARegionEvent regionEventLatitudeIsValid:latitude]) {
            UA_LERR(@"Proximity region latitude must not be greater than %d or less than %d degrees.", kUARegionEventMaxLatitude, kUARegionEventMinLatitude);
            return;
        }

        _latitude = latitude;
    }
}

- (void)setLongitude:(NSNumber *)longitude {
    if (longitude != _longitude) {

        if (longitude && ![UARegionEvent regionEventLongitudeIsValid:longitude]) {
            UA_LERR(@"Proximity region longitude must not be greater than %d or less than %d degrees.", kUARegionEventMaxLongitude, kUARegionEventMinLongitude);
            return;
        }

        _longitude = longitude;
    }
}

- (void)setRSSI:(NSNumber *)RSSI {
    if (RSSI != _RSSI) {

        if (RSSI && ![UARegionEvent regionEventRSSIIsValid:RSSI]) {
            UA_LERR(@"Proximity region RSSI must not be greater than %d or less than %d dBm.", kUAProximityRegionMaxRSSI, kUAProximityRegionMinRSSI);
            return;
        }

        _RSSI = RSSI;
    }
}

- (BOOL)isValid {
    if ((self.latitude && !self.longitude) || (!self.latitude && self.longitude)) {
        UA_LERR(@"A proximity region's latitude and longitude must both be set.");
        return NO;
    }

    if (!self.minor || self.minor.intValue < 0 || self.minor.intValue > UINT16_MAX) {
        UA_LERR(@"Minor cannot be nil, less than zero or greater than 65535.");
        return NO;
    }

    if (!self.major || self.major.intValue < 0 || self.major.intValue > UINT16_MAX) {
        UA_LERR(@"Major cannot be nil, less than zero or greater than 65535.");
        return NO;
    }

    if (![UARegionEvent regionEventCharacterCountIsValid:self.proximityID]) {
        UA_LERR(@"Proximity region ID must not be greater than %d or less than %d characters in length.", kUARegionEventMaxCharacters, kUARegionEventMinCharacters);
        return NO;
    }

    return YES;
}

@end

File: Airship/AirshipCore/Source/UAProximityRegion.swift
/* Copyright Airship and Contributors */

/**
 * A proximity region defines an identifier, major and minor.
 */
@objc
public class UAProximityRegion : NSObject {

    let latitude: Double?
    let longitude: Double?
    let rssi: Double?
    let proximityID: String
    let major: Double
    let minor: Double

    /**
     * Default constructor.
     *
     * @param proximityID The ID of the proximity region.
     * @param major The major.
     * @param minor The minor.
     * @param rssi The rssi.
     * @param latitude The latitude of the circular region's center point in degress.
     * @param longitude The longitude of the circular region's center point in degrees.
     *
     * @return Proximity region object or `nil` if error occurs.
     */
    public init?(proximityID: String,
          major:Double,
          minor: Double,
          rssi: Double? = nil,
          latitude: Double? = nil,
          longitude: Double? = nil) {

        if ((latitude != nil || longitude != nil) && (latitude == nil || longitude == nil)) {
            AirshipLogger.error("Invalid proximity region. Both lat and long must both be defined if one is provied.")
            return nil
        }

        if let latitude = latitude {
            guard UAEventUtils.isValid(latitude: latitude) else {
                return nil
            }
        }

        if let longitude = longitude {
            guard UAEventUtils.isValid(longitude: longitude) else {
                return nil
            }
        }

        if let rssi = rssi {
            guard UAProximityRegion.isValid(rssi: rssi) else {
                return nil
            }
        }

        guard UAProximityRegion.isValid(proximityID: proximityID) else {
            return nil
        }


        guard UAProximityRegion.isValid(major: major) else {
            return nil
        }

        guard UAProximityRegion.isValid(minor: minor) else {
            return nil
        }

        self.proximityID = proximityID;
        self.major = major
        self.minor = minor
        self.rssi = rssi
        self.latitude = latitude
        self.longitude = longitude
    }

    /**
     * Factory method for creating a proximity region.
     *
     * @param proximityID The ID of the proximity region.
     * @param major The major.
     * @param minor The minor.
     *
     * @return Proximity region object or `nil` if error occurs.
     */
    @objc(proximityRegionWithID:major:minor:)
    public class func proximityRegion(proximityID: String,
                                      major: Double,
                                      minor: Double) -> UAProximityRegion?{
        return UAProximityRegion(proximityID: proximityID, major: major, minor: minor)
    }

    /**
     * Factory method for creating a proximity region.
     *
     * @param proximityID The ID of the proximity region.
     * @param major The major.
     * @param minor The minor.
     * @param rssi The rssi.
     *
     * @return Proximity region object or `nil` if error occurs.
     */
    @objc(proximityRegionWithID:major:minor:rssi:)
    public class func proximityRegion(proximityID: String,
                                      major: Double,
                                      minor: Double,
                                      rssi: Double) -> UAProximityRegion? {
        return UAProximityRegion(proximityID: proximityID, major: major, minor: minor, rssi: rssi)
    }


    /**
     * Factory method for creating a proximity region.
     *
     * @param proximityID The ID of the proximity region.
     * @param major The major.
     * @param minor The minor.
     * @param latitude The latitude of the circular region's center point in degress.
     * @param longitude The longitude of the circular region's center point in degrees.
     *
     * @return Proximity region object or `nil` if error occurs.
     */
    @objc(proximityRegionWithID:major:minor:latitude:longitude:)
    public class func proximityRegion(proximityID: String,
                                      major: Double,
                                      minor: Double,
                                      latitude: Double,
                                      longitude: Double) -> UAProximityRegion? {
        return UAProximityRegion(proximityID: proximityID, major: major, minor: minor, latitude: latitude, longitude:longitude)
    }

    /**
     * Factory method for creating a proximity region.
     *
     * @param proximityID The ID of the proximity region.
     * @param major The major.
     * @param minor The minor.
     * @param rssi The rssi.
     * @param latitude The latitude of the circular region's center point in degress.
     * @param longitude The longitude of the circular region's center point in degrees.
     *
     * @return Proximity region object or `nil` if error occurs.
     */
    @objc(proximityRegionWithID:major:minor:rssi:latitude:longitude:)
    public class func proximityRegion(proximityID: String,
                                      major: Double,
                                      minor: Double,
                                      rssi: Double,
                                      latitude: Double,
                                      longitude: Double) -> UAProximityRegion? {
        return UAProximityRegion(proximityID: proximityID, major: major, minor: minor, rssi: rssi, latitude: latitude, longitude:longitude)
    }

    private class func isValid(proximityID: String) -> Bool {
        guard proximityID.count > 0 && proximityID.count <= 255 else {
            AirshipLogger.error("Invalid proximityID \(proximityID). Must be between 1 and 255 characters")
            return false
        }
        return true
    }

    private class func isValid(rssi: Double) -> Bool {
        guard rssi >= -100 && rssi <= 100 else {
            AirshipLogger.error("Invalid RSSI \(rssi). Must be between -100 and 100")
            return false
        }
        return true
    }

    private class func isValid(major: Double) -> Bool {
        guard major >= 0 && major <= 65535 else {
            AirshipLogger.error("Invalid major \(major). Must be between 0 and 65535")
            return false
        }
        return true
    }

    private class func isValid(minor: Double) -> Bool {
        guard minor >= 0 && minor <= 65535 else {
            AirshipLogger.error("Invalid minor \(minor). Must be between 0 and 65535")
            return false
        }
        return true
    }
}
